flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java

```diff
@@ -297,17 +299,22 @@ private static DataType convertToDataType(Schema schema) {
 	 * @return Avro's {@link Schema} matching this logical type.
 	 */
 	public static Schema convertToSchema(LogicalType logicalType) {
-		return convertToSchema(logicalType, "record");
+		return convertToSchema(logicalType, "record", true);
 	}
 
 	/**
 	 * Converts Flink SQL {@link LogicalType} (can be nested) into an Avro schema.
 	 *
 	 * @param logicalType logical type
 	 * @param rowName     the record name
+	 * @param top         whether it is parsing the root record,
+	 *                    if it is, the logical type nullability would be ignored
 	 * @return Avro's {@link Schema} matching this logical type.
 	 */
-	public static Schema convertToSchema(LogicalType logicalType, String rowName) {
+	public static Schema convertToSchema(
+			LogicalType logicalType,
+			String rowName,
+			boolean top) {
```

Comment:
Is it necessary? Shouldn't the `top` be handled in the planner and just passed with a correct setting in the `logicalType`? I think it is a bit too deep to fix it here.

Reply:
It is meaningless to have a nullable top row type, and in the `CREATE TABLE` DDL there is no way to specify that the table row is not nullable, actually it should always be not null(even if all the fields are null).

Reply:
I agree that the top row should not be nullable. It is a property of a SQL Table though, not of the format. If I am not wrong there you can not specify a ROW nullable in `CREATE TABLE` because it is always `NOT NULL`. Therefore it is meaningless to have additional parameter in the `convertToSchema`.

What I am saying is that the format should not be bother with this. It is purely a matter of the planner.

Reply:
I have the same feeling with @dawidwys .

Reply:
Currently, only avro format cares about the outer row nullability, we can switch to change the planner if we found more user cases.

Technically to say, it is impossible to infer the outer row nullability only from the DDL.

Reply:
Sorry, but I don't understand your arguments.

What do you mean by "inferring" the outer row nullability? There is no inference. In SQL the outer row is always `NOT NULL` or otherwise the row simply does not exist.

In the current shape you have two arguments that contradict each other. Take this example:

```
DataType type = DataTypes.ROW(
    DataTypes.FIELD("f0", DataTypes.ROW(
        DataTypes.FIELD("f1", DataTypes.BIGINT()
    ).nullable()
).nullable();

Schema schema = convertToSchema(type, "record", true) // <- you're overriding the property of type
// you can achieve the same with
// Schema schema = convertToSchema(type.notNull(), "record")
```

It's the Planners task to produce a `notNull` type for a Tables type.

Reply:
> `In SQL the outer row is always NOT NULL`
It is not true, a row (null, null) is nullable true. And i don't think it makes sense to change the planner behavior in general in order to fix a specific use case. 

>// you can achieve the same with
>// Schema schema = convertToSchema(type.notNull(), "record") 

I don't think we should let each invoker to decide whether to make the data type not null, because in current codebase, we should always do that, make the decision everyone is error-prone and hard to maintain.

I have merge the `top` into a row type nullability switch, see `AvroSchemaConverter.convertToSchema(LogicalType logicalType)`.

Reply:
> It is not true, a row (null, null) is nullable true. And i don't think it makes sense to change the planner behavior in general in order to fix a specific use case.

Excuse me, but I wholeheartedly disagree with your statement. null =/= (null, null). (null, null) is still `NOT NULL`. A whole row in a Table can not be null. Only particular columns can be null. Therefore the top level row of a Table is always `NOT NULL`.

I am not suggesting changing planner behaviour for a particular use case. The planner should always produce `NOT NULL` type for a top level row of a Table. If it doesn't, it is a bug.

Reply:
> I don't think we should let each invoker to decide whether to make the data type not null, because in current codebase, we should always do that, make the decision everyone is error-prone and hard to maintain.

I agree making the same decision over and over again at multiple location is error prone and hard to maintain and that's what I want to avoid.

Reply:
>Excuse me, but I wholeheartedly disagree with your statement. null =/= (null, null). (null, null) is still NOT NULL. A whole row in a Table can not be null. Only particular columns can be null. Therefore the top level row of a Table is always NOT NULL.

You can test it in PostgreSQL with the following SQL:
```sql
create type my_type as (a int, b varchar(20));

create table t1(
  f0 my_type,
  f1 varchar(20)
);

insert into t1 values((null, null), 'def');

select f0 is null from t1; -- it returns true
```

Reply:
But the `my_type` is not a top level row in your example. It is a type of a column. It has nothing to do with the case we're discussing.

Reply:
I had offline discussion with @wuchong and @dawidwys , and after some research, we found that a non-nullable row type is more reasonable.

But because the change is huge(many codes that convert a type info to data type assumes nullable true before), me and @wuchong decide to change the method signature to `convertToSchema(RowType schema)` and add a notion to the method doc that the passed in `schema` must be the top level record type.

Reply:
@dawidwys  I have changed the schema row type to be always nullable false, please take a look again if you have time, thanks so much.